src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -0,0 +1,331 @@
+      fatal(
+        s"""expected cell type Int32, Int64, Float32, Float64, or String but got:
+           |    ${ cellType.toPrettyString() }
+         """.stripMargin)
+    }
+
+    val sc = hc.sc
+    val hConf = hc.hadoopConf
+
+    val header1 = parseHeader(getHeaderLines(hConf, files.head), sep)
+    val header1Bc = sc.broadcast(header1)
+
+    val sampleIds: Array[String] =
+      if (dropSamples)
+        Array.empty
+      else
+        header1
+
+    val nSamples = sampleIds.length
+
+    LoadMatrix.warnDuplicates(sampleIds)
+
+    val lines = sc.textFilesLines(files, nPartitions.getOrElse(sc.defaultMinPartitions))
+
+    val fileByPartition = lines.partitions.map(p => partitionPath(p))
+    val firstPartitions = fileByPartition.zipWithIndex.filter {
+      case (name, partition) => partition == 0 || fileByPartition(partition - 1) != name
+    }.map { case (_, partition) => partition }.toSet
+
+    val matrixType = MatrixType(VSMMetadata(
+      sSignature = TString(),
+      vSignature = TString(),
+      genotypeSignature = cellType
+    ))
+
+    val keyType = matrixType.kType
+    val rowKeys: RDD[RegionValue] = lines.mapPartitionsWithIndex { (i, it) =>
+      if (firstPartitions(i)) {
+        val hd1 = header1Bc.value
+        val hd = it.next().value.split(sep)
+        if (!hd1.sameElements(hd)) {
+          if (hd1.length != hd.length) {
+            fatal(
+              s"""invalid sample ids: lengths of headers differ.
+                 |    ${ hd1.length } elements in ${ files(0) }
+                 |    ${ hd.length } elements in ${ fileByPartition(i) }
+               """.stripMargin
+            )
+          }
+          hd1.zip(hd).zipWithIndex.foreach { case ((s1, s2), j) =>
+            if (s1 != s2) {
+              fatal(
+                s"""invalid sample ids: expected sample ids to be identical for all inputs. Found different sample ids at position $j.
+                   |    ${ files(0) }: $s1
+                   |    ${ fileByPartition(i) }: $s2""".
+                  stripMargin)
+            }
+          }
+        }
+      }
+
+      val region = MemoryBuffer()
+      val rvb = new RegionValueBuilder(region)
+      val rv = RegionValue(region)
+      it.map { v =>
+        val line = v.value
+        if (line.nonEmpty) {
+          val k = line.substring(0, line.indexOf(sep))
+          region.clear()
+          rvb.start(keyType)
+          rvb.startStruct()
+          rvb.addString(k)
+          rvb.addString(k)
+          rvb.endStruct()
+          rv.setOffset(rvb.end())
+        }
+        rv
+      }
+    }
+
+    val rdd = lines
+      .mapPartitionsWithIndex { (i, it) =>
+        val region = MemoryBuffer()
+        val rvb = new RegionValueBuilder(region)
+        val rv = RegionValue(region, 0)
+
+        if (firstPartitions(i))
+          it.next()
+
+        it.map { v =>
+          val line = v.value
+          if (line.nonEmpty) {
```

Comment:
I think this will  cause an empty line to duplicate the previous line, since you always return a `RegionValue`?  You should filter out empties.  I'm slightly dubious on our empty line checks actually, so otherwise, you can just parse it normally (and generate whatever error you get).